### PR TITLE
Added Utxo table

### PR DESCRIPTION
**Pull request by DavidGershony**
Added Utxo table to allow faster updating of the addresses on input and avoid the heavy merge
this will also be used to calc the rich list much faster

### CODE REVIEW

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 1]
```diff
@@ -838,11 +838,17 @@ public async Task DeleteBlockAsync(string blockHash)
...
+         FilterDefinition<InputTable> inputFilter = Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
+         var inputs = InputTable.FindSync(inputFilter).ToList();
 
-         FilterDefinition<OutputTable> addrForOutputFilter = Builders<OutputTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
-         Task<DeleteResult> output = OutputTable.DeleteManyAsync(addrForOutputFilter);
+         Task utxos = UtxoTable.InsertManyAsync(inputs.Select(_ => new UtxoTable
+            {
+               Address = _.Address, Outpoint = _.Outpoint, Value = _.Value, BLockIndex = _.BlockIndex
+            }))
+            .ContinueWith(task => InputTable.DeleteManyAsync(inputFilter));
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:848]: shouldn't this be async/await?

Comment (dangershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:848]: ho its returning the task right?

Comment (sondreb) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:848]: Yes, without the brackets and one line like this, is equal to { return InputTable... } ... if I'm not reviewing this too early.

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 2]
```diff
@@ -838,11 +838,17 @@ public async Task DeleteBlockAsync(string blockHash)
...
-         FilterDefinition<InputTable> addrForInputFilter = Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
-         Task<DeleteResult> input = InputTable.DeleteManyAsync(addrForInputFilter);
+         FilterDefinition<InputTable> inputFilter = Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
+         var inputs = InputTable.FindSync(inputFilter).ToList();
 
-         FilterDefinition<OutputTable> addrForOutputFilter = Builders<OutputTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
-         Task<DeleteResult> output = OutputTable.DeleteManyAsync(addrForOutputFilter);
+         Task utxos = UtxoTable.InsertManyAsync(inputs.Select(_ => new UtxoTable
+            {
+               Address = _.Address, Outpoint = _.Outpoint, Value = _.Value, BLockIndex = _.BlockIndex
```
Comment (sondreb) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:846]: L is capitalized in BLockIndex.

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoDb.cs [hunk 3]
```diff
@@ -55,6 +55,14 @@ public IMongoCollection<InputTable> InputTable
          }
       }
 
+      public IMongoCollection<UtxoTable> UtxoTable
+      {
+         get
+         {
+            return mongoDatabase.GetCollection<UtxoTable>("Utxo");
```
Comment (sondreb) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoDb.cs:62]: So we have both a collection named "Transaction" and this introduces"Utxo"? How about "UnspentTransaction" instead?

**Review by sondreb — APPROVED** [review]
I did a quick review, I'm not at this time able to determine the quality of most of the LINQ-query logic without running myself. Let me know if that is needed 👍🏼

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs (hunk 1 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:848]: Sorry, this code is not ment to go to master, I'm going to change the logic for the entire delete because we don't have indexes on the tables we delete from

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoDb.cs (hunk 3 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoDb.cs:62]: Well we have output so I'll add unspend Outputs

File: src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+         Builders<AddressComputedTable>.Filter.Eq(addr => addr.ComputedBlockIndex, blockIndex);
+      Task<DeleteResult> addressComputed = storage.AddressComputedTable.DeleteManyAsync(addrCompFilter);
+
+      // delete computed history
+      FilterDefinition<AddressHistoryComputedTable> addrCompHistFilter =
+         Builders<AddressHistoryComputedTable>.Filter.Eq(addr => addr.BlockIndex, blockIndex);
+      Task<DeleteResult> addressHistoryComputed = storage.AddressHistoryComputedTable.DeleteManyAsync(addrCompHistFilter);
+
+      // delete computed utxo
+      FilterDefinition<AddressUtxoComputedTable> addrCompUtxoFilter =
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:64]: We wont need this table anymore right? I guess there will be a separate PR to delete it?

File: src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+using Blockcore.Indexer.Core.Storage.Mongo.Types;
+using MongoDB.Bson;
+using MongoDB.Driver;
+
+namespace Blockcore.Indexer.Core.Storage.Mongo;
+
+public static class BlockRewindOperation
+{
+   public static async Task RewindBlockOnIbdAsync(this MongoData storage, long blockIndex)
+   {
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:15]: this method seems complicated, I know its an edge case but later we defo need tests here.

File: src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+               new FindOneAndDeleteOptions<T>
+               {
+                  Sort = new BsonDocumentSortDefinition<T>(new BsonDocument("_id", -1))
+               });
+         }
+
+      } while (true);
+   }
+
+   private static Task MergeRewindInputsToUnspentTransactionsAsync(MongoData storage ,long blockIndex)
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:147]: I think perhaps such complex methods need a good description, something like:

"Inputs spend outputs, when an output is spent it gets deleted from the UnspendOutput table and the action of the delete is represented in the inputs table, when a rewind happens we need to bring back outputs that have been deleted from the UnspendOutput so we look for those outputs in the inputs table, however the block index in the inputs table is the one representing the input not the output we are trying to restore so we have to look it up in the outputs table."

Comment (dangershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:147]: This got me thinking the inputs we are returning will always be for the current block index no?

File: src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+         .Project(_ => new
+         {
+            //We need the block index that the output was created on, the rest of the data is the same as input
+            Value = _[nameof(InputTable.Value)],
+            Address = _[nameof(InputTable.Address)],
+            BlockIndex = _[output][nameof(OutputTable.BlockIndex)],
+            Outpoint = new
+            {
+               OutputIndex = _[nameof(Outpoint)][nameof(Outpoint.OutputIndex)],
+               TransactionId = _[nameof(Outpoint)][nameof(Outpoint.OutputIndex)],
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:181]: should this be?
`TransactionId = _[nameof(Outpoint)][nameof(Outpoint.TransactionId )]`

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 8]
```diff
@@ -830,39 +831,21 @@ private class MapAddressBag
...
-
-         // delete computed history
-         FilterDefinition<AddressHistoryComputedTable> addrCompHistFilter = Builders<AddressHistoryComputedTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
-         Task<DeleteResult> addressHistoryComputed = AddressHistoryComputedTable.DeleteManyAsync(addrCompHistFilter);
+          SyncBlockInfo block = BlockByHash(blockHash);
 
-         // delete computed utxo
-         FilterDefinition<AddressUtxoComputedTable> addrCompUtxoFilter = Builders<AddressUtxoComputedTable>.Filter.Eq(addr => addr.BlockIndex, block.BlockIndex);
-         Task<DeleteResult> addressUtxoComputed = AddressUtxoComputedTable.DeleteManyAsync(addrCompUtxoFilter);
+          var rewindTask = globalState.IndexModeCompleted
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:842]: I don't think `IndexModeCompleted` is the right flag to use here becuase its only set after the indexer task is run, so if on startup we have uncomplete blocks that will trigger a rewind even if indexes are built.

I think we need to check in mongo if indexes are built instead, or a combination of both options 

```
if(!globalState.IndexModeCompleted)
{
  // call mongo to check if indexes are built.
}
```

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs [hunk 9]
```diff
@@ -62,44 +61,53 @@ public void AddToStorageBatch(StorageBatch storageBatch, SyncBlockTransactionsOp
...
                ScriptOutputInfo res = scriptInterpeter.InterpretScript(syncConnection.Network, output.ScriptPubKey);
-               string addr = res != null ? (res?.Addresses != null && res.Addresses.Any()) ? res.Addresses.First() : res.ScriptType.ToString() : "none";
+               string addr = res != null
+                  ? (res?.Addresses != null && res.Addresses.Any()) ? res.Addresses.First() : res.ScriptType.ToString()
+                  : "none";
 
-               return new OutputTable
+               var outpoint = new Outpoint { TransactionId = trx.GetHash().ToString(), OutputIndex = index };
+
+               storageBatch.OutputTable.AddOrReplace(outpoint.ToString(), new OutputTable
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:74]: Why `AddOrReplace`? is there a chance an output could be there already? seems incorrect, we should throw if that happens to catch it.

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs [hunk 10]
```diff
@@ -126,60 +135,91 @@ public SyncBlockInfo PushStorageBatch(StorageBatch storageBatch)
...
 
-         var t4 = Task.Run(() =>
+         var inputTableTask = Task.CompletedTask;
+         if (storageBatch.InputTable.Any())
          {
-            if (storageBatch.InputTable.Any())
-               data.InputTable.InsertMany(storageBatch.InputTable, new InsertManyOptions {IsOrdered = false});
-         });
+            var utxosLookups = FetchUtxos(storageBatch.InputTable
+               .Where(_ => _.Address == null)
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:158]: I wonder couldn't this could be filtered in the mongo call (just an idea for improvement)

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs [hunk 11]
```diff
@@ -126,60 +135,91 @@ public SyncBlockInfo PushStorageBatch(StorageBatch storageBatch)
...
+               .Select(_ => _.Outpoint));
 
-         var t5 = Task.Run(() =>
+            foreach (InputTable input in storageBatch.InputTable)
+            {
+               if (input.Address != null) continue;
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:163]: should we consider throwing here? it means something went wrong no?

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs [hunk 12]
```diff
@@ -126,60 +135,91 @@ public SyncBlockInfo PushStorageBatch(StorageBatch storageBatch)
...
+         var outpointsFromNewInput = storageBatch.InputTable
+            .Select(_ => _.Outpoint)
+            .ToList();
+
+         if (outpointsFromNewInput.Any())
+         {
+            var filterToDelete = Builders<UnspentOutputTable>.Filter
+               .Where(_ => outpointsFromNewInput.Contains(_.Outpoint));
+
+            data.UnspentOutputTable.DeleteMany(filterToDelete);
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:221]: In theory you dont have to wait for all tasks to finish just `unspentOutputTableTask` before you can trigger this one right?

```
Task.WaitAll(unspentOutputTableTask)

data.UnspentOutputTable.DeleteMany..
```

File: src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs [hunk 13]
```diff
@@ -196,10 +204,10 @@ await mongoData.InputTable.Indexes
 
                }).ContinueWith(async task =>
                {
-                  log.LogDebug($"Creating indexes on {nameof(InputTable)}.{nameof(InputTable.Address)}");
+                  log.LogDebug($"Creating indexes on {nameof(UnspentOutputTable)}.{nameof(UnspentOutputTable.Address)}");
 
-                  await mongoData.InputTable.Indexes
-                     .CreateOneAsync(new CreateIndexModel<InputTable>(Builders<InputTable>
+                  await mongoData.UnspentOutputTable.Indexes
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs:209]: Aren't you missing in index on the `BlockIndex` on the `UnspentOutputTable`

File: src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs [hunk 14]
```diff
@@ -378,7 +386,7 @@ await mongoData.Mempool.Indexes
             new BsonDocument("$lookup",
                new BsonDocument
                {
-                  { "from", "Output" },
+                  { "from", "UnspentOutput" },
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs:389]: We don't need all this code anymore why not delete it?

File: src/Blockcore.Indexer/appsettings.json [hunk 15]
```diff
@@ -41,6 +41,6 @@
       "AverageInterval": 10,
 
       // Store the trx hex in mongo storage or read it from RPC
-      "StoreRawTransactions": true
+      "StoreRawTransactions": false
```
Comment (dangershony) [inline src/Blockcore.Indexer/appsettings.json:44]: This we have to remember to flip on production deployment, @sondreb FYI David proposes we will not store trx data in store by default but read it form the client, IMO for prod deployments that the indexer will use we should store them in mongo

File: src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs (hunk 7 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs:181]: Yes, I'd rather use the class name than magic strings in the call to mongo

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs (hunk 8 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:842]: Don't we have a way to indicate to all tasks that the state has chagned from IDB?

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs (hunk 9 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:74]: That was for the bitcoin issue that was fixed ages ago, can't remember the BIP number but I can add a todo there for the hard coding of the 2 transactions that have the same output like they did in BTC

File: src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs (hunk 13 above)
Comment (DavidGershony) [reply src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs:209]: It is being created with the creation of the table

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs (hunk 9 above)
Comment (dangershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs:74]: ah yeah right, yes add a todo.

File: src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs [hunk 16]
```diff
@@ -276,142 +284,13 @@ await mongoData.Mempool.Indexes
...
-                  }
-
-                  totalSecondsToMerge += totalSeconds;
-
-                  return true;
-               }
-               else
-               {
-                  Runner.GlobalState.IndexMode = false;
-                  Runner.GlobalState.IndexModeCompleted = true;
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Sync/SyncTasks/BlockIndexer.cs:330]: I think you deleted this by mistake

**Review by dangershony — APPROVED** [review]

File: src/Blockcore.Indexer/appsettings.json (hunk 15 above)
Comment (sondreb) [reply src/Blockcore.Indexer/appsettings.json:44]: What does it mean "by default"? You mean during development and running locally on developer computer?

This setting is set by the published chain info JSON file, so we'll just make sure all JSON files there has this set to true.
